staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go

```diff
@@ -305,6 +309,22 @@ func (qs *queueSet) StartRequest(ctx context.Context, hashValue uint64, flowDist
 	return req, false
 }
 
+func getRequestWaitContext(ctx context.Context) (context.Context, context.CancelFunc) {
+	deadline, ok := ctx.Deadline()
+	if !ok {
+		return ctx, func() {}
+	}
+
+	// we will allow the request to wait in the queue for one third of the time
+	// of its remaining deadline.
+	timeout := time.Until(deadline) / 3
```

Comment:
(1) The remaining time needs to be according to the queueset's clock.  Thus, the request's arrival time also needs to be according to the queueset's clock.  In normal usage, that is the real clock.  In testing, that is the clock used in the test.

(2) Let's stick with the current quantity of one quarter of the request duration limit.